src/enumerable.cr

```diff
@@ -1205,6 +1205,50 @@ module Enumerable(T)
     reduce(initial) { |memo, e| memo * (yield e) }
   end
 
+  # Returns an array of tuples containing the cartesian product
+  # between the collection and the enumerable. Or in other words, it
+  # will return an Array with a Tuple for each possible combination of
+  # two elements from the collection and the passed enumerable.
+  #
+  # ```
+  # arr = [1, 2, 3]
+  # set = Set(Char).new(['a', 'b'])
+  # arr.product(set)
+  # => [{1, 'a'}, {1, 'b'}, {2, 'a'}, {2, 'b'}, {3, 'a'}, {3, 'b'}])
+  # ```
+  #
+  # An empty array will be returned if either collection is empty.
+  def product(ary : Enumerable(U)) forall U
+    result = if ary.responds_to?(:size)
```

Comment:
This is not quite correct. `Enumerable` also has a `size` method, that merely iterates through all of the elements to count them. So this will traverse the enumerable many times.

I think `product` is fine in `Array`, and shouldn't be available in `Enumerable`.

We'll eventually revisit all of the std design, in one go, instead of deciding the fate of each method case by case. I don't think I'll be able to merge this PR or take any other decision about this right now.